java/compaction/compaction-job-execution/src/main/java/sleeper/compaction/job/execution/CompactionJobMessageHandler.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2022-2024 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sleeper.compaction.job.execution;
+
+import com.amazonaws.services.sqs.model.Message;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import sleeper.compaction.job.CompactionJob;
+import sleeper.configuration.properties.instance.InstanceProperties;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static sleeper.configuration.properties.instance.CompactionProperty.COMPACTION_TASK_MAX_CONSECUTIVE_FAILURES;
+import static sleeper.configuration.properties.instance.CompactionProperty.COMPACTION_TASK_MAX_TIME_IN_SECONDS;
+
+public class CompactionJobMessageHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CompactionJobMessageHandler.class);
+
+    private final Supplier<Instant> timeSupplier;
+    private final int maxConsecutiveFailures;
+    private final int maxTimeInSeconds;
+    private final MessageReceiver messageReceiver;
+    private final MessageConsumer messageConsumer;
+    private final FailedJobHandler failedJobHandler;
+
+    public CompactionJobMessageHandler(InstanceProperties instanceProperties, Supplier<Instant> timeSupplier,
+            MessageReceiver messageReceiver, MessageConsumer messageConsumer, FailedJobHandler failedJobHandler) {
+        maxTimeInSeconds = instanceProperties.getInt(COMPACTION_TASK_MAX_TIME_IN_SECONDS);
+        maxConsecutiveFailures = instanceProperties.getInt(COMPACTION_TASK_MAX_CONSECUTIVE_FAILURES);
+        this.timeSupplier = timeSupplier;
+        this.messageReceiver = messageReceiver;
+        this.messageConsumer = messageConsumer;
+        this.failedJobHandler = failedJobHandler;
+    }
+
+    public Result runAt(Instant startTime) throws InterruptedException, IOException {
+        Instant maxTime = startTime.plus(Duration.ofSeconds(maxTimeInSeconds));
+        int numConsecutiveFailures = 0;
+        Instant currentTime = startTime;
+        long totalNumberOfMessagesProcessed = 0;
+        while (numConsecutiveFailures < maxConsecutiveFailures && currentTime.isBefore(maxTime)) {
+            Optional<JobAndMessage> jobAndMessageOpt = messageReceiver.receiveMessage();
+            if (!jobAndMessageOpt.isPresent()) {
+                numConsecutiveFailures++;
```

Comment:
It seems like the reason to track the consecutive failures would be to stop it repeatedly failing to process a job if there's a problem with the task? Incrementing the number of consecutive failures here seems like it would be more to do with how long the task has been idle, which can be handled separately now. Can we remove this?